### ISSUE TITLE
Disable cache

### ISSUE THREAD

**Issue by kappy** [JimBobSquarePants/ImageProcessor#155]
Hi

How can I disable cache entirely? I want the Cdn to be the only one doing cache upon first request of an image. The web app will only serve images from a blob.

I don't wan't to clutter disk with the diskcache or either a cache container on azure. I also don't the the 302 Found response to the cached image.

Is this possible via configuration?

**Comment by kappy** [comment]
Update: setting the maxDays="-1" returns the following error:
"The value for the property 'maxDays' is not valid. The error is: The value must be inside the range 0-2147483647."

**Comment by kappy** [comment]
Update: setting the maxDays="0" does serve out the images on the fly, but it creates the empty folder paths on the filesystem (using DiskCache)

**Comment by JimBobSquarePants** [comment]
You could create your own implementation of an `IImageCache` which always returns true for expired. There's information in the documentation [here](http://imageprocessor.org/imageprocessor-web/extending/#iimagecache) that will tell you a little more about that.

What CDN are you using? WIll it cache querystrings? I have a plugin for Azure blob storage already written.

**Comment by kappy** [comment]
I'm using Azure Cdn and it supports query string.

I'm using that plugin, but it will always serve the images by redirecting the the cache blob container or the Cdn. 

I wan't the inverse process. I want to serve the link direly from Cdn, falling back to the app when the Cdn doesn't have the image.

And the most important, not leaving out trash like empty folders and files on the filesystem.

The current approach that you have is nice if you only think on client side caching, but the process is not optimized for server side caching. the 302 responses take a lot of time.
Caching should not be your primary concern because there are other alternatives that already do that way better. At least, the cache should be possible to disable.

Imagine a Huge (and I really mean HUGE) site with a lot of traffic and images. having them all requested by the website is not ideal. you should be able to point them to Cdn and then falling back to the site if not available, in order to reduce the amount of requests the site will have to respond.

**Comment by kappy** [comment]
I would also want to make a suggestion, parallel to the /remote.axd handler, having a virtual path providers, directly tighten to certain storage.

I would prefer to call my images by "/myStorage/image.png" instead of "/remove.axd/mystorage..blob.core.windows.net/mystorage/image.png"

Again, for Cdn links, this would be a great improvement
"az00000000.vo.msecnd.net/remove.axd/mystorage..blob.core.windows.net/mystorage/image.png" its already huge

**Comment by JimBobSquarePants** [comment]
Like I said. Implement your own cache and you can disable it. 

To be honest, my experience with Azure etc is very limited. I wrote that plugin in two days having never seen the SDK before. How would the CDN know to revert to the app instead of returning a 404? If the library is not installed on the CDN then it can't know about the image request. 

If you wrote your own IImageService implementation you could prefix the value with whatever you wanted. I only use remote.axd as default as it doesn't interfere with routing.

I'm happy to take on any improvements you can offer as long as I can keep things backwards compatible. That's been the biggest problem for me with this library so far.... 

Building this has been a long and intense learning experience so as you can imagine, there are lots of things in the library I would like to improve. I never, ever thought it would prove to be as popular as it has become.

**Comment by kappy** [comment]
The way of Pull Cdn's work, is that if the requested resource doesn't exist on the cache, it will internally request the origin url (configuration setting) and serve that content while save it on cache for next request.

**Comment by JimBobSquarePants** [comment]
Ahhh ok.... Well there a plans afoot for a major rewrite of both the core and web libraries so we will have scope to break a few things and rewrite them if need be.

**Comment by kappy** [comment]
You gave me enough hints to try build the scenario I want. I'll do it prob next Week (going away next days) and let you know the results

**Comment by JimBobSquarePants** [comment]
Excellent. Best of luck! :smile:

**Comment by kappy** [comment]
Check this out.
https://gist.github.com/kappy/1f07d13cbc720b423574

This was the minimum code to work "without" cache. 
Indeed you need a better design. It should not the the cache the responsibility of outputting to the browser. As much, it should intercept the earliest, to override the request to the cached object.
In cases of cache miss, then it should generate images, store it, and serve the image, in separate. But that's just my tough ;)

**Comment by JimBobSquarePants** [comment]
I'm not sure I follow what you are saying. If it's that I should not instruct the browser to cache the image then that is incorrect. Every image processing library that is in production instructs the browser to cache the image.

**Comment by kappy** [comment]
I wasn't saying that. The "cache" entity should only store or retrieve the content, it should not handle the output/rewrite of the http context. that should be the module job.
It's even 1 less dependency that way.

To me, you have 2 design flaws can be improved:

 * https://gist.github.com/kappy/1f07d13cbc720b423574#file-nullcache-L14 => you need the set CachedPath or else every request will end up with a YSOD. Is that is a requirement, it should be handled on the base class as part of the process.
* https://gist.github.com/kappy/1f07d13cbc720b423574#file-nullcache-L30 => this is the only place where the response is set. It also relies on having a virtual path for it and http redirects. This should not be on the cache. If much, the module should only ask the cache for the path (the early it can), and either redirect or output the content itself. I don't know.. I might be wrong.. just thinking out loud

In short... I was expecting a simple `return true` on `IsNewOrUpdatedAsync` to disable the cache and still work as intended, but that proved me wrong.

**Comment by JimBobSquarePants** [comment]
Different cache providers require different information to correctly serve the image and there is no way to correctly predict what is needed. That is why they rewrite the path.

**Comment by JimBobSquarePants** [comment]
I've added functionality to the AzureBlobCache version [1.0.4](https://www.nuget.org/packages/ImageProcessor.Web.Plugins.AzureBlobCache/1.0.4) that allows streaming the file rather than rewriting. Additionally I've added a new `CloudImageService` to ImageProcessor.Web which lets you set better urls to your blob location.

There were two other issues #161 #178 which were asking for the reverse CDN functionality also and these updates solve their issues so that should let you do what you need to.

I've added the extra configuration to the docs to show how to switch to the new behaviour. 

http://imageprocessor.org/imageprocessor-web/plugins/azure-blob-cache/#config
http://imageprocessor.org/imageprocessor-web/configuration/#securityconfig

Hopefully that clears everything up.

**Comment by kappy** [comment]
I think so too. Awesome. will check that later.

**Comment by JimBobSquarePants** [comment]
Use ImageSharp

**Comment by wtct** [comment]
> 
> 
> Use ImageSharp

I'm using ImageProcessor integrated with Umbraco so it can be tricky to swap to ImageSharp. I'm looking for any example of disabling cache in ImageProcessor but I need to stop overal caching (creating directories and cached files) not only to stop serving images from cache.

**Comment by JimBobSquarePants** [comment]
Disabling the cache is DDOS waiting to happen. You do not want to do that.

**Comment by wtct** [comment]
It all depends on the hosting environment. We have two proxy servers with caching requests like this. So caching in ImageProcessor is waste of performance due to ImageProcessor is waiting for cached file saved and then transmit it to user.

**Comment by JimBobSquarePants** [comment]
For the amount of customisation you’re talking about you’d be better off writing your own implementation. As you would have seen in the readme this library is in maintenance mode. It’s feature complete and will not be updated.

**Comment by JimBobSquarePants** [comment]
@wtct I’ve just realised that we have previously discussed disabling the ImageProcessor.Web cache with you in another issue where you were given a clear and concise answer. Spamming our issue tracker like this and wasting our time is incredibly rude and I will have to block you if you ever do it again. Consider this a severe  warning. 

P.S. if your Umbraco solution requires 2 caching proxy servers you have developmental quality issues that cannot be solved here.

**Comment by wtct** [comment]
> 
> 
> @wtct I’ve just realised that we have previously discussed disabling the ImageProcessor.Web cache with you in another issue where you were given a clear and concise answer. Spamming our issue tracker like this and wasting our time is incredibly rude and I will have to block you if you ever do it again. Consider this a severe warning.
> 
> P.S. if your Umbraco solution requires 2 caching proxy servers you have developmental quality issues that cannot be solved here.

In previous issue only received information about implementing IImageCache possibility but I had to drop this topic due to other more important tasks. However currently I'm working with images and decided to go deeply with disabling cache in ImageProcessor.

In this issue is linked example with custom IImageCache but it doesn't work any more. I wanted to make sure my assumptions.

Unfortunately any suggestions obout implementing own IImageCache are useless at the end for what I need to do because I need to disable overal caching purpose. I have to stop creating files and folders because it is not secure, efficient and useless in my configuration. IImageCache force RewritePath method implementation which have to rewrite path for previously created one. So it is impossible to disable this behavior.

Really it is big shame that Umbraco decided to use ImageProcessor. You mentioned about DDoS possibility but it doesn't matter if you have disabled or enabled caching. You can overload any Umbraco site by a lot of image requests with different size parameters.

Two caching proxies helps a lot and I can't imagine hight traffic website without that. You have no idea about my site and proxy so don't judge, please.

**Comment by JimBobSquarePants** [comment]
>Really it is big shame that Umbraco decided to use ImageProcessor. 

Note to others. This is the kind of comment that will get you immediately blocked from all my repositories. 

>You mentioned about DDoS possibility but it doesn't matter if you have disabled or enabled caching. You can overload any Umbraco site by a lot of image requests with different size parameters.

This is also nonsense. It's been possible for a VERY long time to be able to restrict resize operations to a specific allowable set of dimensions.
